File: engine-common/src/main/java/com/sos/scheduler/engine/common/system/OperatingSystem.java
package com.sos.scheduler.engine.common.system;

import com.google.common.base.Joiner;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;

import static com.google.common.base.Strings.nullToEmpty;
import static com.google.common.collect.ObjectArrays.concat;

public abstract class OperatingSystem {
    public static final String name = System.getProperty("os.name");
    public static final CpuArchitecture cpuArchitecture = CpuArchitecture.cpuArchitecture();
    public static final boolean isWindows = name.startsWith("Windows");
    public static final boolean isUnix = !isWindows;
    public static final Unix unix = new Unix();
    public static final Windows windows = new Windows();
    public static final OperatingSystem operatingSystem = isWindows? windows : unix;
    public static final String javaLibraryPathPropertyName = "java.library.path";
    private static final Logger logger = LoggerFactory.getLogger(OperatingSystem.class);
    
    public final String makeModuleFilename(String path) {
        File file = new File(path);
        return new File(file.getParent(), System.mapLibraryName(file.getName())).getPath();
    }
    
    public abstract String makeExecutableFilename(String name);
    public abstract String getDynamicLibraryEnvironmentVariableName();

    public String hostname() {
//        try {
//            return java.net.InetAddress.getLocalHost().getCanonicalHostName();
//        } catch (Exception e) {
//            logger.trace(e.toString(), e);
            return alternativeHostname();
//        }
    }

    protected abstract String alternativeHostname();

    public static class Windows extends OperatingSystem {
        @Override public final String makeExecutableFilename(String name) {
            return name + ".exe";
        }

        @Override public final String getDynamicLibraryEnvironmentVariableName() {
            return "PATH";
        }

        protected final String alternativeHostname() {
            return nullToEmpty(System.getenv("COMPUTERNAME"));
        }

//        @Override public void removeAbsoluteDirectoryRecursivly(File directory) {
//            executeShell(commandFile(), "rd", "/s", directory.toString());
//        }
//
//        private File commandFile() {
//            return new File(windowsDirectory() + File.separator + "system32", "cmd.exe");
//        }
//
//        private File windowsDirectory() {
//            String windir = System.getenv("WINDIR");
//            if (isNullOrEmpty(windir))  throw new RuntimeException("Environment variable WINDIR is missing");
//            File result = new File(windir);
//            assert result.isDirectory();
//            return result;
//        }
    }
    

    public static class Unix extends OperatingSystem {
        @Override public final String makeExecutableFilename(String name) {
            return name;
        }

        @Override public final String getDynamicLibraryEnvironmentVariableName() {
            return "LD_LIBRARY_PATH";
        }

        protected final String alternativeHostname() {
            return nullToEmpty(System.getenv("HOSTNAME"));
        }
//        @Override public void removeAbsoluteDirectoryRecursivly(File directory) {
//            executeShell(new File("/bin/rm"), "-rf", directory.toString());
//        }
    }

    public static void prependJavaLibraryPath(File f) {
        String a = System.getProperty(javaLibraryPathPropertyName);
        String c = concatFileAndPathChain(f, nullToEmpty(a));
        if (!c.equals(a)) {
            System.setProperty(javaLibraryPathPropertyName, c);
            logger.debug("Property " + javaLibraryPathPropertyName + "=" + c);
        }
    }

    public static String concatFileAndPathChain(File f, String pathChain) {
        String abs = f.getAbsolutePath();
        String[] b = pathChain.split(File.pathSeparator);
        for (int i = 0; i < b.length; i++)  if (b[i].isEmpty() || b[i].equals(abs))  b[i] = null;
        return Joiner.on(File.pathSeparatorChar).skipNulls().join(concat(abs, b));
    }
}

File: engine-common/src/main/scala/com/sos/scheduler/engine/common/system/OperatingSystem.scala
package com.sos.scheduler.engine.common.system

import com.sos.scheduler.engine.common.scalautil.AutoClosing.autoClosing
import com.sos.scheduler.engine.common.scalautil.FileUtils.implicits._
import java.io.{File, FileInputStream}
import java.nio.file.Files.newDirectoryStream
import java.nio.file.{Path, Paths}
import scala.io
import scala.util.Try

/**
 * @author Joacim Zschimmer
 */
trait OperatingSystem {

  def makeExecutableFilename(name: String): String

  def getDynamicLibraryEnvironmentVariableName: String

  def hostname: String = alternativeHostname

  def distributionNameAndVersionOption: Option[String]

  def cpuModel: Option[String]

  protected def alternativeHostname: String
}

object OperatingSystem {
  val name: String = sys.props("os.name")
  val cpuArchitecture = CpuArchitecture.cpuArchitecture
  val isWindows = name startsWith "Windows"
  val isUnix = !isWindows
  val isSolaris = name startsWith "SunOS"
  lazy val unix = new Unix
  lazy val windows = new Windows
  val operatingSystem: OperatingSystem = if (isWindows) windows else unix
  val javaLibraryPathPropertyName = "java.library.path"
  lazy val KernelSupportsNestedShebang = KernelVersion() >= KernelVersion("Linux", List(2, 6, 28))  // Exactly 2.6.27.9 - but what means the fourth number? http://www.in-ulm.de/~mascheck/various/shebang/#interpreter-script

  def makeModuleFilename(path: String): String = {
    val file = new File(path)
    new File(file.getParent, System.mapLibraryName(file.getFileName.toString)).toString
  }

  def makeExecutableFilename(name: String): String = operatingSystem.makeExecutableFilename(name)

  def getDynamicLibraryEnvironmentVariableName: String = operatingSystem.getDynamicLibraryEnvironmentVariableName

  final class Windows private[system] extends OperatingSystem {
    def makeExecutableFilename(name: String): String = name + ".exe"

    def getDynamicLibraryEnvironmentVariableName: String = "PATH"

    protected def alternativeHostname: String = sys.env.getOrElse("COMPUTERNAME", "")

    def distributionNameAndVersionOption = None

    def cpuModel = sys.env.get("PROCESSOR_IDENTIFIER")
  }

  final class Unix private[system] extends OperatingSystem {
    def makeExecutableFilename(name: String): String = name

    def getDynamicLibraryEnvironmentVariableName: String = "LD_LIBRARY_PATH"

    protected def alternativeHostname: String = sys.env.getOrElse("HOSTNAME", "")

    lazy val distributionNameAndVersionOption: Option[String] = {
      def readFirstLine(file: Path): String =
        autoClosing(new FileInputStream(file)) { in ⇒
          io.Source.fromInputStream(in).getLines.next().trim
        }

      def readFileOsRelease() = {
        val prettyNamePrefix = "PRETTY_NAME="
        val file = "/etc/os-release"
        autoClosing(new FileInputStream(file)) { in ⇒    // http://man7.org/linux/man-pages/man5/os-release.5.html
          io.Source.fromInputStream(in).getLines collectFirst {
            case line if line startsWith prettyNamePrefix ⇒
              line.stripPrefix(prettyNamePrefix).stripPrefix("\"").stripPrefix("'").stripSuffix("\"").stripSuffix("'").trim
          }
        }
        .getOrElse { sys.error(s"Key PRETTY_NAME not found in file $file")}
      }

      def readFileAnyRelease() = {
        val anyFile = autoClosing(newDirectoryStream(Paths.get("/etc"), "*-release")) { stream ⇒
          val iterator = stream.iterator
          if (!iterator.hasNext) throw sys.error("No file matches /etc/*-release")
          iterator.next
        }
        readFirstLine(anyFile)
      }

      Try { readFirstLine(Paths.get("/etc/system-release")) }  // Best result under CentOS 7.2 (more version details than in /etc/os-release)
        .recover { case _ ⇒ readFileOsRelease() }   // New standard ?
        .recover { case _ ⇒ readFileAnyRelease() }  // Vendor-specific
        .recover { case _ ⇒ readFirstLine(Paths.get("/etc/release")) } // Solaris ?
        .toOption
    }

    def cpuModel =
      Try {
        val CpuModelRegex = """model name[ \t]*:[ \t]*(.+)""".r
        autoClosing(new FileInputStream("/proc/cpuinfo")) { in ⇒
          io.Source.fromInputStream(in).getLines collectFirst {  // Assuming all cores are of same model
            case CpuModelRegex(model) ⇒ model.trim
          }
        }
      } .toOption.flatten
  }

  def concatFileAndPathChain(f: File, pathChain: String): String = {
    val abs = f.getAbsolutePath
    abs +: (pathChain split File.pathSeparator filter { o ⇒ o.nonEmpty && o != abs }) mkString File.pathSeparatorChar.toString
  }
}
